**Let's continue from question 4 in our notebook setup in PySpark**

### 4. Using pyspark, read the data you copied into HDFS in Step 3. You may either use Jupyterhub on EMR (the default user and password are jovyan and jupyter) or work from pyspark in the terminal if you prefer. Once you have created a pyspark DataFrame, complete the following steps below:

In [2]:
df = spark.read.csv('s3://brainstation-dsft/eng_1M_1gram.csv', header=True)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [3]:
# Let's take a peak at our data
df.show(10)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+---------+----+---------+-----+-----+
|    token|year|frequency|pages|books|
+---------+----+---------+-----+-----+
|inGermany|1927|        2|    2|    2|
|inGermany|1929|        1|    1|    1|
|inGermany|1930|        1|    1|    1|
|inGermany|1933|        1|    1|    1|
|inGermany|1934|        1|    1|    1|
|inGermany|1935|        1|    1|    1|
|inGermany|1938|        5|    5|    5|
|inGermany|1939|        1|    1|    1|
|inGermany|1940|        1|    1|    1|
|inGermany|1942|        2|    2|    2|
+---------+----+---------+-----+-----+
only showing top 10 rows

#### 4.1 Show the schema of the table in pyspark

In [4]:
# This shows the schema of the dataframe
df.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- token: string (nullable = true)
 |-- year: string (nullable = true)
 |-- frequency: string (nullable = true)
 |-- pages: string (nullable = true)
 |-- books: string (nullable = true)

#### 4.2 Display the total number of rows of data

In [5]:
df.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

261823225

The count() feature displays the rows in our dataframe, we have ~261 Million rows (261823225)

#### 4.3 Create a new DataFrame from a query using Spark SQL, filtering out only rows where the token is "data"

Let's convert our process into Spark.sql which makes it easier to work with and write codes in SQL.

In [6]:
# Let's create our view in Spark session
df.createOrReplaceTempView("1gram")

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [9]:
# Let's select rows which has 'data' in the token column
filtered= spark.sql("SELECT * FROM 1gram WHERE token= 'data'")

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [7]:
# Let's check if our token column has 'data' in it
spark.sql("SELECT * FROM 1gram WHERE token= 'data'").show(3)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+-----+----+---------+-----+-----+
|token|year|frequency|pages|books|
+-----+----+---------+-----+-----+
| data|1584|       16|   14|    1|
| data|1614|        3|    2|    1|
| data|1627|        1|    1|    1|
+-----+----+---------+-----+-----+
only showing top 3 rows

### 4.4 Get a count of the number of rows for the above.

In [11]:
# Now let's get a count of the rows
filtered.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

316

Here we have 316 rows with 'data' in the token column.

### 4.5 Write the filtered data back to a directory in the Hadoop filesystem from Spark using df.write.csv() - e.g. /user/hadoop/eng_data_1gram. Be sure to pass the header=True parameter. Examine the contents of what you've written using hadoop fs -ls.

Let's put our filtered dataset in our HDFS with .write.csv()

In [15]:
filtered.write.option("header","true").csv("/user/hadoop/eng_data_1gram")

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

### 5. Collect the contents of the directory into a single file on the local drive of the head node using getmerge: e.g. hadoop fs -getmerge /user/hadoop/eng_data_1gram eng_data_1gram.csv. Move this file into a S3 bucket in your account using aws s3 cp.

In [ ]:
#hadoop fs -getmerge /user/hadoop/eng_data_1gram eng_data_1gram.csv